### DISCUSSION TITLE
Monorepos - please explain why a root eslintrc for each workspace.

### DISCUSSION THREAD

**Discussion by dylang** [AndreaPontrandolfo/sheriff#61] [Q&A]
> While Sheriff can technically be used at the root of monorepos, it's not recommended.
> It works best when applied to individual packages within a monorepo.

* Can you explain why? 
* _What_ works best? 
* What are drawbacks of just having one root config?

I have a monorepo with hundreds of packages. Should we have hundreds of eslint configs? And if we need an override for, say, a React rule, we should repeat that override hundreds of times?

Great looking project, thanks for putting so much effort into the code and documentation.

**Comment by AndreaPontrandolfo** [comment]
Hey @dylang!
I'm happy to have a chat about this, because it's **definitely** something i've been meaning to explore more in depth.
So, this recommendation is more than anything based on my personal needs and anecdotal experience. 
For all my projects (both personal and professional) i've used Sheriff this way, and paired with the [eslint.workingdirectories](https://github.com/microsoft/vscode-eslint#mono-repository-setup) setting, it seems to offer an optimal in-editor experience. 

That being said, i think i should better explore the difference in performance about both approaches and inform the user about it in the docs.

> And if we need an override for, say, a React rule, we should repeat that override hundreds of times?

Well, the rationale behind the currently recommended approach is: every package is supposed to have different needs and so configuring Sheriff independently in every package is actually a benefit intended by design. 
**BUT**, if you want to share some configs in every package of your monorepo, you can very easily do so.
You can create something like a `constants` workspace package and export a variable like `myUniversalConfigSlice`, then in every other package you import this `constants` package and in the `eslint.config.js` you can append the object in the config, right after Sheriff. This should work.

So, i'll keep exploring this topic and i'll update this discussion accordingly.  The backlog is long though, so it might take a while.
I'm eager to also hear your next insights on the topic as it evolves, if you want to share them :)

```js
// eslint.config.js

import sheriff from "eslint-config-sheriff";
import { defineFlatConfig } from "eslint-define-config";
import { myUniversalConfigSlice } from "@constants"

  const sheriffOptions = {
    react: false,
    next: false,
    lodash: false,
    playwright: false,
    jest: false,
    vitest: false,
  };

  export default defineFlatConfig([...sheriff(sheriffOptions), myUniversalConfigSlice]);
```

**Comment by dylang** [comment]
> every package is supposed to have different needs

I've found the `files` filter helpful for this, because our different needs vary by filename, not by directory. For example:

 * `**/*.test.*` - unit tests have jest and testing library specific rules. 
 * `**/*.browser.*` - browser tests have some unique rules.
 * `**/*.js` - needs a couple things that typescript normally handles.

**Comment by AndreaPontrandolfo** [comment]
Relevant discussion, we should keep tracking: https://github.com/eslint/eslint/discussions/16960

**Reply by lishaduck** [reply]
These days, it's 100% recommended to place eslint.config.js in the root directory.
If there's not a technical reason on sheriff's side, would you be up for a PR to update the docs to reflect that?

P.S.: Thanks for Sheriff! It's helped @PSDTools immensely over the last year or so.

**Reply by AndreaPontrandolfo** [reply]
> P.S.: Thanks for Sheriff! It's helped https://github.com/PSDTools immensely over the last year or so.

That's great to hear, i'm glad!

> These days, it's 100% recommended to place eslint.config.js in the root directory.

Are you referring to something specific? Do you have a link?

> If there's not a technical reason on sheriff's side

- not Sheriff-specific: i think stuff like `eslint-plugin-import` wouldn't work well with that setup.
- Sheriff-specific: the upside to defining a `eslint.config.js`at the every workspace level, is that then you can customize the SheriffOptions per workspace, which is usually very handy.

**Reply by lishaduck** [reply]
> not Sheriff-specific: i think stuff like eslint-plugin-import wouldn't work well with that setup.

Yeah, that would make sense.

> Sheriff-specific: the upside to defining a eslint.config.jsat the every workspace level, is that then you can customize the SheriffOptions per workspace, which is usually very handy.


eslint/eslint#17853 (the main one)
eslint/eslint#16960 (you've already mentioned this one)

Nicholas has very much recommended putting one eslint config in the root, and then using files to filter it down to just the packages that care.
Generally, I think that makes sense, but it does require some changes to sheriff to allow specifying it. (See below)

> Yeah, unfortunately this is going to be heavier than it was previously, but also, it will be a lot less error-prone. The cascading nature of eslintrc meant that complexity exploded exponentially with each additional config file in the cascade. The way we typically suggest people address a use case like yours (and I'm not sure this will work for you in a monorepo, but has been our advice) is to **have a single config file at the root of your project and then use files to target specific configurations at specific directories.** So instead of a core config file and then directory-specific override files, just have one root config file that specifies the directory-level overrides. — Nicholas Zakas [Emphasis added]

I think I've seen it a few other spots, and I haven't read the docs recently, but I belive it said so there as well.

Here's the api I'd imagine:

<details><summary>Idea 1</summary>
<p>

```ts
import { sheriff } from "eslint-config-sheriff";
import { defineFlatConfig } from "eslint-define-config";

export default defineFlatConfig([
  ...sheriff({
    files: ["packages/foo/*.{ts,tsx}"],
    ignores: ["**/*.d.ts"],
    languageOptions: {
      ecmaVersion: 2022,
    },
    linterOptions: {},
    plugins: {},
    processor: {},
    settings: {},
    presets: {
      react: false,
      lodash: false,
      next: false,
      playwright: false,
      jest: false,
      vitest: false,
      astro: false,
    },
    rules: {
      "@typescript-eslint/naming-convention": "off",
      "prefer-destructuring": "off",
    },
  }),
]);
```

</p>
</details> 

The benefits being that I can, i.e., configure all of the projects in `apps/` with one config, and then those two projects in another place.

Alternatively,

<details><summary>Idea 2</summary>
<p>

```ts
import { sheriff } from "eslint-config-sheriff";
import { defineFlatConfig } from "eslint-define-config";

export default defineFlatConfig([
  ...sheriff({
    files: ["packages/foo/*.{ts,tsx}"],
    ignores: ["**/*.d.ts"],
    presets: {
      react: false,
      lodash: false,
      next: false,
      playwright: false,
      jest: false,
      vitest: false,
      astro: false,
    },
  }),
  {
    files: ["packages/foo/*.{ts,tsx}"],
    ignores: ["**/*.d.ts"],
    languageOptions: {
      ecmaVersion: 2022,
    },
    linterOptions: {},
    plugins: {},
    processor: {},
    settings: {},
    rules: {
      "@typescript-eslint/naming-convention": "off",
      "prefer-destructuring": "off",
    },
  },
]);
```

</p>
</details> 

A bit of duplication, but more explicit about what's going on, and supports more customization with less work.

**Reply by AndreaPontrandolfo** [reply]
@lishaduck The "idea 2" is already totally possible today, Sheriff doesn't impede it in any way, but we discourage it in the docs.

**Reply by AndreaPontrandolfo** [reply]
about "idea 1", i think i don't understand what the benefit should be.

**Reply by lishaduck** [reply]
Yeah, as I was writing it, I was surprised to find intellisense for 2.
For 1, it would just be a way of writing a block like normal, but sheriff would inject the sheriff rules alongside.
As I've thought about it, 2 makes the most sense to me.

I guess my question at this point is why idea 2 is discouraged if it also allows customizing per workspace.

**Reply by AndreaPontrandolfo** [reply]
I'll definitely compile a pros and cons checklist soon. 
Anyway, another thing that crossed my mind, when you use create-sheriff-config, if you don't point  the process to a specific workspace, it won't automatically infer the correct defaults for the Sheriffs options.

**Reply by lishaduck** [reply]
Oh, I realized I forgot to reply here:

> Anyway, another thing that crossed my mind, when you use create-sheriff-config, if you don't point the process to a specific workspace, it won't automatically infer the correct defaults for the Sheriffs options.

But that's not a limitation of anything external, if we desired, we could support it (it'd just require more work).

**Reply by lishaduck** [reply]
Two more points:

Pro: [`eslint-plugin-turbo`](https://www.npmjs.com/package/eslint-plugin-turbo) support
Cons (sorta): eslint/eslint#18742 improves the situation (behind a flag until v10,[^1] and I think supporting running it from the root is still useful)

[^1]: Nicholas said v10 will likely be before the year is out though in the comments of the RFC.
